src/offset/local/windows.rs

```diff
@@ -8,132 +8,195 @@
+    let tz_info = match tz_info_for_year(utc.year()) {
+        Some(tz_info) => tz_info,
+        None => return LocalResult::None,
+    };
+    LocalResult::Single(match (tz_info.std_transition, tz_info.dst_transition) {
+        (Some(std_transition), Some(dst_transition)) => {
+            let std_transition_utc = std_transition - tz_info.dst_offset;
+            let dst_transition_utc = dst_transition - tz_info.std_offset;
+            if dst_transition_utc < std_transition_utc {
+                if utc >= &dst_transition_utc && utc < &std_transition_utc {
+                    tz_info.dst_offset
+                } else {
+                    tz_info.std_offset
+                }
+            } else {
+                if utc >= &std_transition_utc && utc < &dst_transition_utc {
+                    tz_info.std_offset
+                } else {
+                    tz_info.dst_offset
+                }
+            }
+        }
+        _ => tz_info.std_offset,
+    })
 }
 
+// We don't use `TzSpecificLocalTimeToSystemTime` because it doesn't let us choose how to handle
+// ambiguous cases (during a DST transition). Instead we get the timezone information for the
+// current year and compute it ourselves, like we do on Unix.
 pub(super) fn offset_from_local_datetime(local: &NaiveDateTime) -> LocalResult<FixedOffset> {
-    offset(local, true)
-}
-
-/// Converts a local `NaiveDateTime` to the `time::Timespec`.
-pub(super) fn offset(d: &NaiveDateTime, local: bool) -> LocalResult<FixedOffset> {
-    let naive_sys_time = system_time_from_naive_date_time(d);
-
-    let local_sys_time = match local {
-        false => from_utc_time(naive_sys_time),
-        true => from_local_time(naive_sys_time),
+    let tz_info = match tz_info_for_year(local.year()) {
+        Some(tz_info) => tz_info,
+        None => return LocalResult::None,
     };
-
-    if let Ok(offset) = local_sys_time {
-        return LocalResult::Single(offset);
+    match (tz_info.std_transition, tz_info.dst_transition) {
+        (Some(_), Some(_)) => tz_info.lookup_with_dst_transitions(*local),
+        _ => LocalResult::Single(tz_info.std_offset),
     }
-    LocalResult::None
-}
-
-fn from_utc_time(utc_time: SYSTEMTIME) -> Result<FixedOffset, Error> {
-    let local_time = utc_to_local_time(&utc_time)?;
-    let utc_secs = system_time_as_unix_seconds(&utc_time)?;
-    let local_secs = system_time_as_unix_seconds(&local_time)?;
-    let offset = (local_secs - utc_secs) as i32;
-    Ok(FixedOffset::east_opt(offset).unwrap())
 }
 
-fn from_local_time(local_time: SYSTEMTIME) -> Result<FixedOffset, Error> {
-    let utc_time = local_to_utc_time(&local_time)?;
-    let utc_secs = system_time_as_unix_seconds(&utc_time)?;
-    let local_secs = system_time_as_unix_seconds(&local_time)?;
-    let offset = (local_secs - utc_secs) as i32;
-    Ok(FixedOffset::east_opt(offset).unwrap())
+// The basis for Windows timezone and DST support has been in place since Windows 2000. It does not
```

Comment:
All these APIs use the same underlying data so have the similar limitations. The data is stored in the registry; the structure of which is documented [here](https://learn.microsoft.com/en-us/windows/win32/api/timezoneapi/ns-timezoneapi-time_zone_information#remarks) and [here](https://learn.microsoft.com/en-us/windows/win32/api/timezoneapi/ns-timezoneapi-dynamic_time_zone_information#remarks). I can't find documentation on the 2004 limitation but that appears to be accurate based on my system (though it is possible to manually edit the registry to add more).

Modern versions of Windows do have [icu.dll](https://learn.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-) which (amongst over things) contains ICU timezone data exposed via C APIs. This might be worth adding to Chrono at some point but even if so then they'll still need to be a fallback so that would be a whole other PR (imho).

Reply:
Great info @ChrisDenton .  I suggest placing your reply into the code as a code comment including the links.

Reply:
> Do we have references for this? Does this PR make our timezone support worse than what we had before? Seems like a pretty limited implementation...

It is very limited, yes. But it is what the `Local` time on Windows PCs is. We are not getting worse in any way, just using the same underlying data as @ChrisDenton says.

I did have some more references but have misplaced them...

Note that the details here don't matter all that much. `GetTimeZoneInformationForYear` just provides a couple of offsets and transition dates for a particular year, and we have to look up where a datetime fails between the transitions.